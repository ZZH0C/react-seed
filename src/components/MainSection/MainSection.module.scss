.hidden{
  display: none;
}

.main_section{
  position: relative;
}