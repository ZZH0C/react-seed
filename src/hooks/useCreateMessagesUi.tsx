import React, { useContext, useEffect } from 'react';
import { useGetMessages } from './useGetMessages';
import { GoogleMessage } from '../models/GoogleMessage';
import { MessageItem } from '../components/MessageItem/MessageItem';
import { sortMessageData } from './useSortMessageData';
import { UserContext } from '../pages/HomePage/HomePage';

export const useCreateMessagesUi = () => {
  const userData = useContext(UserContext);
  const { setMessageList, state } = useGetMessages();

  const messages: JSX.Element[] = [];

  useEffect(() => {
    if (userData && 'accessToken' in userData) {
      const token = userData.accessToken;
      setMessageList(token);
    }
    if (!userData) {
      setMessageList(null);
    }
  }, [userData]);
  if (state.length > 0) {
    state.forEach((e: GoogleMessage) => {
      const messageData = sortMessageData(e.value);
      messages.push(
        <MessageItem
          key={Math.random()}
          fromWho={messageData.from}
          messageSnippet={messageData.snippet}
          messageTitle={messageData.title}
          messageDate={messageData.date}
          messageId={e.value.id}
        />,
      );
    });
  }
  return messages;
};
