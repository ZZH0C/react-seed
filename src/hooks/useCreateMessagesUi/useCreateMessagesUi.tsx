import React, { useCallback, useContext, useEffect, useState } from 'react';
import { Direction, useGetMessages } from '../useGetMessages/useGetMessages';
import { GoogleMessagePromise } from '../../models/GoogleMessage';
import { MessageItem } from '../../components/MessageItem/MessageItem';
import { sortMessageData } from '../useSortMessageData/useSortMessageData';
import { UserContext } from '../../pages/HomePage/HomePage';
import { useQueryParams } from '../useQueryParams/useQueryParams';
import { useLocation } from 'react-router-dom';
import { PaginationButton } from '../../components/PaginationButton/PaginationButton';
import map from 'lodash/map';
import { Loader } from '../../components/Loader/Loader';
import styles from '../../components/MainSection/MainSection.module.scss';
import classNames from 'classnames';

export const useCreateMessagesUi = (): JSX.Element => {
  const userData = useContext(UserContext);
  const { setMessageList, state, clearMessageList } = useGetMessages();
  const { getGoogleQueryParams } = useQueryParams();
  const location = useLocation();
  const googleData = getGoogleQueryParams();
  let token: string | null = null;
  if (userData && 'accessToken' in userData) {
    token = userData.accessToken;
  }

  const [loaderState, setLoaderState] = useState(false);

  const activateLoader = useCallback(() => {
    setLoaderState(true);
  }, []);

  const disableLoader = useCallback(() => {
    setLoaderState(false);
  }, []);

  const refreshPage = useCallback(() => {
    activateLoader();
    setMessageList(token, googleData, Direction.current, disableLoader);
  }, [googleData, setMessageList, token]);

  useEffect(() => {
    activateLoader();
    clearMessageList();
    if (userData && 'accessToken' in userData) {
      refreshPage();
    }
    if (!userData) {
      setMessageList(token, '', Direction.current, disableLoader);
    }
    //TODO: add pagination and userData?.accessToken
  }, [userData, location.search]);

  const createMessage = (message: GoogleMessagePromise) => {
    const messageData = sortMessageData(message.value);
    return (
      <MessageItem
        key={Math.random()}
        fromWho={messageData.from}
        messageSnippet={messageData.snippet}
        messageTitle={messageData.title}
        messageDate={messageData.date}
        messageId={message.value.id}
        refreshPageCallback={refreshPage}
        activateLoader={activateLoader}
      />
    );
  };
  return (
    <>
      <div className={classNames(styles.main_section)}>
        <Loader isActive={loaderState} />
        {state.messages.length > 0 ? map(state.messages, createMessage) : []}
      </div>
      <div className="controlButtons">
        <PaginationButton
          isDisabled={state.pages.length < 3}
          isRight={false}
          onClickFunction={() => {
            activateLoader();
            setMessageList(token, googleData, Direction.prev, disableLoader);
          }}
        >
          Previous
        </PaginationButton>
        <PaginationButton
          isDisabled={
            state.pages[state.pages.length - 1] === undefined ||
            state.pages.length < 2
          }
          isRight={true}
          onClickFunction={() => {
            activateLoader();
            setMessageList(token, googleData, Direction.next, disableLoader);
          }}
        >
          Next
        </PaginationButton>
      </div>
    </>
  );
};
