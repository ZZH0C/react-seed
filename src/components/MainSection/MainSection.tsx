import React, { useContext, useEffect } from 'react';
import { UserContext } from '../../pages/HomePage/HomePage';
import { useGetMessages } from '../../hooks/useGetMessages';
type MainSectionProps = React.HTMLAttributes<HTMLElement>;

export const MainSection: React.FC<MainSectionProps> = () => {
  const userData = useContext(UserContext);
  const { setMessageList, state } = useGetMessages();
  // const [state, setState] = useState([]);

  const messages: any[] = [];

  useEffect(() => {
    if (userData !== null) {
      if ('accessToken' in userData) {
        const token = userData.accessToken;
        setMessageList(token);
        console.log(state);

        // setState(QQ);
      }
    }
  }, [userData]);

  return <section className="mainSection">{messages}</section>;
};
