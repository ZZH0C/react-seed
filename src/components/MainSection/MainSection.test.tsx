import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
import { MainSection } from './MainSection';
import { mocked } from 'ts-jest/utils';
import { useCreateMessagesUi } from '../../hooks/useCreateMessagesUi/useCreateMessagesUi';

jest.mock('../../hooks/useCreateMessagesUi/useCreateMessagesUi');
const mockElement: JSX.Element = (
  <div>
    <p>MOCK_ELEMENT</p>
  </div>
);
mocked(useCreateMessagesUi).mockImplementation(() => mockElement);

describe('components/ButtonLink', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render and match snapshot', () => {
    expect(shallow(<MainSection />)).toMatchSnapshot();
  });
});
