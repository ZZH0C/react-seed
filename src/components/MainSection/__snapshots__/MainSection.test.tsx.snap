// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ButtonLink should render and match snapshot 1`] = `
<section
  className="mainSection"
>
  <SearchBar />
  <div>
    <p>
      MOCK_ELEMENT
    </p>
  </div>
</section>
`;
